### PR TITLE
Prevent unintentional exceptions constituting a passing test.

### PR DESCRIPTION
**Pull request by martica**
This add a default test to assertCodeExecution that no exception was thrown and not caught within the test code, and fixes the many places where this was happening.

There is an allow_exceptions parameter to use for cases where the test is specifically testing that exceptions can be thrown and not caught.

This is a solution to #792.

## PR Checklist:

- [x] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: tests/utils.py [hunk 1]
```diff
@@ -681,10 +702,8 @@ def makeTempDir(self):
         except FileExistsError:
             pass
 
-    def runAsJavaScript(
-                self, main_code, extra_code=None, js=None,
-                run_in_function=False, args=None, python_exists=False
-            ):
+    def runAsJavaScript(self, main_code, extra_code=None, js=None, run_in_function=False, wrap_in_try_catch=False,
```
Comment (brutusthebee) [inline tests/utils.py]: At column 5: (C901) 'TranspileTestCase.runAsJavaScript' is too complex (16)

File: tests/datatypes/test_str.py [hunk 2]
```diff
@@ -1034,23 +1055,27 @@ def test_with_tuple_of_dicts(self, js_cleaner, py_cleaner):
...
+                    print('format this: %(arg)s' % ({'arg':'spam'}, {'arg2':'eggs'}))
+                except TypeError as e:
+                    print(e)
                 print('Done.')
                 """)
 
-            self.assertCodeExecution(tests, js_cleaner = js_cleaner, py_cleaner = py_cleaner)
+            self.assertCodeExecution(tests, js_cleaner=js_cleaner, py_cleaner=py_cleaner)
 
+        @unittest.expectedFailure
```
Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 9: (F811) redefinition of unused 'test_with_kwargs' from line 1030

File: tests/datatypes/test_str.py [hunk 3]
```diff
@@ -1006,6 +1027,7 @@ def test_bogus_specifier(self, js_cleaner, py_cleaner):
 
             self.assertCodeExecution(test, js_cleaner = js_cleaner, py_cleaner = py_cleaner)
```
Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 54: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 56: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 79: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 81: (E251) unexpected spaces around keyword / parameter equals

File: tests/datatypes/test_str.py [hunk 4]
```diff
@@ -1006,6 +1027,7 @@ def test_bogus_specifier(self, js_cleaner, py_cleaner):
...
 
+        @unittest.expectedFailure
         @transforms(
             js_bool = False,
```
Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 20: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 22: (E251) unexpected spaces around keyword / parameter equals

File: tests/datatypes/test_str.py [hunk 5]
```diff
@@ -1006,6 +1027,7 @@ def test_bogus_specifier(self, js_cleaner, py_cleaner):
...
             decimal = False,
```
Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 20: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 22: (E251) unexpected spaces around keyword / parameter equals

File: tests/datatypes/test_str.py [hunk 6]
```diff
@@ -1034,62 +1056,59 @@ def test_with_tuple_of_dicts(self, js_cleaner, py_cleaner):
...
                 print(">>> '%*(spam)s' % {'spam': 'eggs'}")
-                print('%*(spam)s' % {'spam': 'eggs'})
+                try:
+                    print('%*(spam)s' % {'spam': 'eggs'})
+                except TypeError as e:
+                    print(e)
                 print('Done.')
                 """)
 
             self.assertCodeExecution(tests, js_cleaner = js_cleaner, py_cleaner = py_cleaner)
```
Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 55: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 57: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 80: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 82: (E251) unexpected spaces around keyword / parameter equals

File: tests/datatypes/test_str.py [hunk 7]
```diff
@@ -1034,62 +1056,59 @@ def test_with_tuple_of_dicts(self, js_cleaner, py_cleaner):
...
                 print(">>> 'format this: %(1)s' % {'spam':'eggs'}")
-                print('format this: %(1)s' % {'spam':'eggs'})
+                try:
+                    print('format this: %(1)s' % {'spam':'eggs'})
+                except KeyError as e:
+                    print(e)
                 print('Done.')
                 """)
 
             self.assertCodeExecution(tests, js_cleaner = js_cleaner, py_cleaner = py_cleaner)
```
Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 55: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 57: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 80: (E251) unexpected spaces around keyword / parameter equals

Comment (brutusthebee) [inline tests/datatypes/test_str.py]: At column 82: (E251) unexpected spaces around keyword / parameter equals

**Comment by martica** [comment]
Good call. I think the size of utils.py made me want some more organization. I didn't file an issue, but i did put breaking up utils.py on my own list earlier this week. :)

**Pull request merged**